### DISCUSSION TITLE
Limit the possible runner operation systems ( OS ) for GitHub runners within a organization

### DISCUSSION THREAD

**Discussion by dgteixeira** [community/community#30512] [Actions]
# Feature Request

## Description
Hello,

This is more of a `Feature Request`, because I've already opened a Support Ticket and they said this is not possible.

What I want to do is, at an organization level, **limit which OS are available for GitHub Actions** that run on github runners.
Also, if possible, select which repositories can be the exception for this rule and for which runner OSs.

## Example / Use case
I want to make sure no repo within an organization can use `macOS` runners and only a set of specific repositories can use `windows` runners and all repos can use `ubuntu` runners.

## Possible Workarounds

Within individual repositories, a potential solution for preventing the use of labels other than `ubuntu-latest` is with a [CODEOWNERS file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners), for example by either:

1. Creating a `CODEOWNERS` file that matches the `.github/workflows` path, so that the code owner must approve any change to a workflow file, OR
2. Creating a new workflow that will automatically check any changes to other workflows, validating that only the desired OS type(s) are used in the files, and then creating a `CODEOWNERS` file to require approval to modify the new validation workflow.

Option `1.` is easier to implement, however more difficult to manage as the code owner will need to approve any change to a workflow file.

Option `2.` will be slightly more complicated to implement (writing a new workflow for this purpose), however will be easier to manage.

## Conclusion

The workarounds are not optimal because they require manual work (approve from `CODEOWNERS`) or the creation of a custom workflow to keep tabs of every repository (and we have to add more repositories as they are created, which is not operable for a big company with thousands of repos).

This way, I would like to request the possibility of **org-wide limitation of the OS on GitHub runners** (and even improve to a repo-level exclusion).

Best regards,
Diogo Teixeira

EDIT: Grammar and typing errors.

**Comment by Akash1134** [comment]
Hi @dgteixeira , Thanks for sharing such detailed post. I have labelled this thread with the relevant tag so our product engineers can review it.

**Reply by dgteixeira** [reply]
Ok, thanks a lot for the help! :)

**Comment by naikrovek** [comment]
can i ask why you want to impose this limitation?  i’m curious what your motivations are, here.

**Reply by dgteixeira** [reply]
Sure!
Recently we went over the GitHub packaged minutes and had to setup additional billing information for the remaining minutes.

This happened because people within the organization started using GitHub Actions on `macOS-latest` believing that it would be the same as `ubuntu-latest`. 
They were using workflows that took over 1h and ran multiple times a day. And they actually didn't need `macOS`, they just wanted hardware acceleration for faster runtime. Multiply that by the 10x `macOS` minutes "cost" in GitHub Actions, and we were out of minutes **really fast**, for everyone in the organization.

This way, we wanted to keep the `ubuntu-latest` for everyone to use and abuse, but keep the _higher rated_ workflows to specific workflows that the teams actually require it. 

Also, we need to understand and prepare for who is going to consume minutes for `macOs` so that their department, within our company, is responsible to pay those amounts (charge back). I know that this can be done after, through the consumption report, but it's something that should be avoided.

EDIT: Grammar and typos

**Reply by naikrovek** [reply]
Ok, that's what I thought you were going to say.  That makes sense.

I don't know of any way to accomplish this, other than to educate your developers and to become very vocal and active with the chargebacks (which is a form of education, followed by financial pain).

**Reply by dgteixeira** [reply]
Yeah, we are using this event like a `Post Mortem` type of ceremony to try and educate everyone, but it's a hard, and long, path to trail.
And protecting our organization from the events of possible 1 or 2 repositories (even if they are well-intended), seems like a good ideia to me.

And, through repo exclusions, the organization admins could allow specific ones to use all the OSs, just like we add repositories to a GitHub App, for example. 

I hope that explained the reasoning behind opening this community discussion (and thanks a lot for taking the time to participate!) :)

**Reply by naikrovek** [reply]
it would be nice if the Windows, Linux, and MacOS runners were in different runner groups, so that this kind of thing would be easier.

**Comment by nebuk89** [comment]
@dgteixeira this is an awesome write up thank you! I have been working on some bits around improving control over access to GitHub's general concurrency.
Just to play back part of your reasoning here :) the main reason you grouped this by 'platform' as how you would like to control access was the cost multiplier per platform correct? 

Was this your only motivator for limiting the OS access or were there other reasons I have not grasped?

**Reply by dgteixeira** [reply]
Hey @nebuk89 , how are you? :)

Yeah, the main reasoning was due to the cost multiplier per platform.
So far, we have not found any other reason for limiting OS access of GitHub runners within our organization.

Basically, if people mess around with `ubuntu-latest`, it does not generally affect the rest of the organization because we have more than enough minutes (if we only use `ubuntu-latest`). 
However, if they do the same with `macOS`, like it happened, it consumed almost every minute real fast.
And when this GitHub minute-budget drains out, the entire organization suffers if we don't have the extra subscription setup and then we have to justify the extra costs.

**Reply by nebuk89** [reply]
Awesome thank you for the feedback, we are experimenting with some designs for UX changes around this :) let me know if you would be interested in providing some feedback :D drop me an email on nebuk89@github.com
